File: src/main/java/com/company/erp/features/chats/service/ChatCallService.java
package com.company.erp.features.chats.service;

import com.company.erp.core.config.AppProperties;
import com.company.erp.core.database.PageQuery;
import com.company.erp.core.exceptions.BadRequestException;
import com.company.erp.core.exceptions.NotFoundException;
import com.company.erp.features.chats.dto.StartCallRequest;
import com.company.erp.features.chats.entity.*;
import com.company.erp.features.chats.presence.PresenceService;
import com.company.erp.features.chats.presence.PresenceStatus;
import com.company.erp.features.chats.repository.ChatCallParticipantRepository;
import com.company.erp.features.chats.repository.ChatCallRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@Transactional
public class ChatCallService {

    private static final Logger log = LoggerFactory.getLogger(ChatCallService.class);

    private static final Set<CallStatus> OPEN_CALL_STATUSES =
            Set.of(CallStatus.RINGING, CallStatus.ANSWERED);
    private static final Set<ParticipantStatus> OPEN_PARTICIPANT_STATUSES =
            Set.of(ParticipantStatus.RINGING, ParticipantStatus.ANSWERED);

    private final ChatCallRepository calls;
    private final ChatCallParticipantRepository participants;
    private final ConversationService conversations;
    private final PresenceService presence;
    private final StreamTokenService streamTokens;
    private final StreamVideoService streamVideo;
    private final AppProperties props;

    public ChatCallService(ChatCallRepository calls,
                           ChatCallParticipantRepository participants,
                           ConversationService conversations,
                           PresenceService presence,
                           StreamTokenService streamTokens,
                           StreamVideoService streamVideo,
                           AppProperties props) {
        this.calls = calls;
        this.participants = participants;
        this.conversations = conversations;
        this.presence = presence;
        this.streamTokens = streamTokens;
        this.streamVideo = streamVideo;
        this.props = props;
    }

    private long ringTimeoutSeconds() {
        return (props.chat() == null || props.chat().call() == null
                || props.chat().call().ringTimeoutSeconds() <= 0)
                ? 60L
                : props.chat().call().ringTimeoutSeconds();
    }

    private long acceptGraceSeconds() {
        return (props.chat() == null || props.chat().call() == null
                || props.chat().call().acceptGraceSeconds() < 0)
                ? 5L
                : props.chat().call().acceptGraceSeconds();
    }

    @Transactional(readOnly = true)
    public ChatCall getById(Long callId) {
        return calls.findWithParticipantsById(callId)
                .orElseThrow(() -> new NotFoundException("Call not found"));
    }

    @Transactional(readOnly = true)
    public Page<ChatCall> historyForUser(Long userId, PageQuery query) {
        return calls.findAllForUser(userId,
                query.toPageable(Set.of("startedAt"), Sort.by(Sort.Direction.DESC, "startedAt")));
    }

    @Transactional(readOnly = true)
    public Page<ChatCall> historyForConversation(Long convId, Long userId, PageQuery query) {
        conversations.requireMember(convId, userId);
        return calls.findByConversationIdOrderByStartedAtDesc(convId,
                query.toPageable(Set.of("startedAt"), Sort.by(Sort.Direction.DESC, "startedAt")));
    }

    public ChatCall start(Long convId, Long callerId, StartCallRequest req) {
        conversations.requireMember(convId, callerId);
        if (calls.existsActiveCallForUser(callerId, OPEN_CALL_STATUSES, OPEN_PARTICIPANT_STATUSES)) {
            throw new BadRequestException("Caller is already in an active call");
        }

        ChatCall c = new ChatCall();
        c.setConversationId(convId);
        c.setCallerId(callerId);
        c.setType(req.type());
        c.setStatus(CallStatus.RINGING);
        c.setStartedAt(Instant.now());
        calls.save(c);
        // Now that the row has an id, stamp the Stream Video CID so every
        // participant joins the same Stream call.
        c.setStreamCallCid(streamTokens.cidForCall(c.getId()));

        Set<Long> memberIds = conversations.memberUserIds(convId);
        for (Long uid : memberIds) {
            ChatCallParticipant p = new ChatCallParticipant();
            p.setId(new ChatCallParticipantId(c.getId(), uid));
            p.setCall(c);
            if (uid.equals(callerId)) {
                p.setStatus(ParticipantStatus.ANSWERED);
                p.setJoinedAt(Instant.now());
            } else {
                p.setStatus(ParticipantStatus.RINGING);
            }
            participants.save(p);
        }
        // Caller is busy from the moment the call starts.
        presence.markBusy(callerId);
        // Ring the callees server-side via Stream's VoIP push (the native
        // CallKit / ConnectionService incoming-call screen) — but ONLY for
        // callees who are OFFLINE: no live STOMP session, which on iOS means
        // the app is backgrounded or killed (the Flutter client drops its
        // WebSocket via `disconnectForBackground` the moment it backgrounds).
        //
        // An ONLINE callee is foreground with the WebSocket up, so it already
        // received the STOMP `call.invite` (broadcast in ChatCallController)
        // and shows the IN-APP incoming-call overlay. Sending that foreground
        // device a VoIP push too is exactly what produced the duplicate native
        // ring header AND the contested audio session (CallKit + WebRTC both
        // grabbing AVAudioSession → mute/unmute yields silence). Gating the
        // ring at the source removes the foreground CallKit ring entirely —
        // no flash — and leaves the foreground audio session to WebRTC alone.
        // OFFLINE callees still get the VoIP push (their only way to ring).
        // Async + exception-swallowing, so a Stream hiccup never breaks
        // call signalling.
        Set<Long> ringTargets = memberIds.stream()
                .filter(uid -> !uid.equals(callerId))
                .filter(uid -> presence.statusOf(uid) == PresenceStatus.OFFLINE)
                .collect(Collectors.toCollection(LinkedHashSet::new));
        if (ringTargets.isEmpty()) {
            log.info("[call] callId={} — all callees ONLINE; skipping Stream VoIP ring "
                    + "(in-app overlay handles foreground, no CallKit)", c.getId());
        } else {
            // Include the caller so Stream makes them the creator (creators are
            // NOT rung); only the OFFLINE callees in this set receive the push.
            Set<Long> ringMembers = new LinkedHashSet<>(ringTargets);
            ringMembers.add(callerId);
            log.info("[call] callId={} — ringing OFFLINE callees via VoIP push: {}",
                    c.getId(), ringTargets);
            streamVideo.ring(c.getStreamCallCid(), callerId, ringMembers);
        }
        // Reload via the EntityGraph so the caller can read c.getParticipants()
        // after the @Transactional boundary closes.
        return calls.findWithParticipantsById(c.getId())
                .orElseThrow(() -> new IllegalStateException("Just-created call not found"));
    }

    public ChatCall accept(Long callId, Long userId) {
        ChatCall c = getById(callId);

        // Grace-window revival — if the sweeper just marked it MISSED but the
        // accept arrives within `acceptGraceSeconds`, restore RINGING and proceed
        // so a late-by-a-few-seconds FCM accept doesn't get rejected.
        if (c.getStatus() == CallStatus.MISSED) {
            long ageSec = Duration.between(c.getStartedAt(), Instant.now()).getSeconds();
            long graceCutoff = ringTimeoutSeconds() + acceptGraceSeconds();
            if (ageSec <= graceCutoff) {
                log.info("[call] REVIVE callId={} accepter user={} ageSec={} graceCutoff={}",
                        callId, userId, ageSec, graceCutoff);
                c.setStatus(CallStatus.RINGING);
                c.setEndedAt(null);
                c.setEndReason(null);
                c.setDurationSeconds(null);
                ChatCallParticipant me = participants.findByCall_IdAndId_UserId(callId, userId)
                        .orElseThrow(() -> new NotFoundException("You are not a participant in this call"));
                if (me.getStatus() == ParticipantStatus.MISSED) {
                    me.setStatus(ParticipantStatus.RINGING);
                    me.setLeftAt(null);
                }
            } else {
                throw new BadRequestException("Call already ended");
            }
        } else if (c.getStatus() != CallStatus.RINGING && c.getStatus() != CallStatus.ANSWERED) {
            throw new BadRequestException("Call already ended");
        }

        ChatCallParticipant p = participants.findByCall_IdAndId_UserId(callId, userId)
                .orElseThrow(() -> new NotFoundException("You are not a participant in this call"));
        if (p.getStatus() != ParticipantStatus.RINGING) {
            return c;
        }
        p.setStatus(ParticipantStatus.ANSWERED);
        p.setJoinedAt(Instant.now());
        if (c.getStatus() == CallStatus.RINGING) {
            c.setStatus(CallStatus.ANSWERED);
            c.setAnsweredAt(Instant.now());
        }
        presence.markBusy(userId);
        return c;
    }

    public ChatCall reject(Long callId, Long userId, String reason) {
        ChatCall c = getById(callId);
        ChatCallParticipant p = participants.findByCall_IdAndId_UserId(callId, userId)
                .orElseThrow(() -> new NotFoundException("You are not a participant in this call"));
        if (p.getStatus() == ParticipantStatus.RINGING) {
            p.setStatus(ParticipantStatus.REJECTED);
            p.setLeftAt(Instant.now());
        }
        // If everyone else has rejected/left, the call ends.
        boolean anyoneActive = c.getParticipants().stream()
                .anyMatch(x -> !x.getUserId().equals(c.getCallerId())
                        && (x.getStatus() == ParticipantStatus.RINGING
                            || x.getStatus() == ParticipantStatus.ANSWERED));
        if (!anyoneActive && c.getStatus() == CallStatus.RINGING) {
            endCallInternal(c, c.getStatus() == CallStatus.RINGING
                    ? CallStatus.REJECTED
                    : CallStatus.ENDED, reason != null ? reason : "rejected");
        }
        return c;
    }

    public ChatCall hangup(Long callId, Long userId) {
        ChatCall c = getById(callId);
        ChatCallParticipant p = participants.findByCall_IdAndId_UserId(callId, userId)
                .orElseThrow(() -> new NotFoundException("You are not a participant in this call"));
        if (p.getStatus() == ParticipantStatus.ANSWERED || p.getStatus() == ParticipantStatus.RINGING) {
            p.setStatus(ParticipantStatus.LEFT);
            p.setLeftAt(Instant.now());
        }

        // Caller leaving ends the call for everyone (Slice 10.2.10 in the guide).
        if (userId.equals(c.getCallerId())) {
            endCallInternal(c, CallStatus.ENDED, "caller_left");
            return c;
        }

        // Otherwise: if no other callee is still in, end it.
        boolean anyoneStillActive = c.getParticipants().stream()
                .filter(x -> !x.getUserId().equals(c.getCallerId()))
                .anyMatch(x -> x.getStatus() == ParticipantStatus.RINGING
                            || x.getStatus() == ParticipantStatus.ANSWERED);
        if (!anyoneStillActive && c.getStatus() != CallStatus.ENDED) {
            endCallInternal(c, CallStatus.ENDED, "all_callees_left");
        } else {
            // Call continues; the user who left is no longer BUSY.
            presence.clearBusy(userId);
        }
        return c;
    }

    /**
     * Sweep every RINGING call older than the configured ring timeout and
     * transition it to MISSED. Called by {@code CallTimeoutScheduler}. Returns
     * the freshly-ended calls so the scheduler can fan out STOMP + FCM
     * notifications to the caller and unanswered callees.
     */
    public List<ChatCall> sweepStaleRinging() {
        Instant cutoff = Instant.now().minus(Duration.ofSeconds(ringTimeoutSeconds()));
        List<ChatCall> stale = calls.findStaleRinging(cutoff);
        List<ChatCall> ended = new ArrayList<>();
        for (ChatCall c : stale) {
            for (ChatCallParticipant p : c.getParticipants()) {
                if (p.getStatus() == ParticipantStatus.RINGING) {
                    p.setStatus(ParticipantStatus.MISSED);
                    p.setLeftAt(Instant.now());
                }
            }
            endCallInternal(c, CallStatus.MISSED, "no_answer");
            ended.add(c);
            log.info("[call] AUTO-MISSED callId={} after {}s (timeout={}s)",
                    c.getId(),
                    Duration.between(c.getStartedAt(), Instant.now()).getSeconds(),
                    ringTimeoutSeconds());
        }
        return ended;
    }

    private void endCallInternal(ChatCall c, CallStatus status, String reason) {
        c.setStatus(status);
        c.setEndedAt(Instant.now());
        c.setEndReason(reason);
        if (c.getAnsweredAt() != null) {
            c.setDurationSeconds((int) Duration.between(c.getAnsweredAt(), c.getEndedAt()).getSeconds());
        } else {
            c.setDurationSeconds(0);
        }
        // Clear BUSY for everyone who was active in this call.
        for (ChatCallParticipant p : c.getParticipants()) {
            if (p.getStatus() == ParticipantStatus.ANSWERED
                    || p.getStatus() == ParticipantStatus.LEFT) {
                presence.clearBusy(p.getUserId());
            }
        }
        // Caller, even if they never "answered", was busy from start.
        presence.clearBusy(c.getCallerId());

        // Cancel the Stream ring for every member. This is the single
        // chokepoint for EVERY terminal end (caller_left, all_callees_left,
        // rejected, no_answer/timeout), so a still-ringing callee that got the
        // ring via Stream's VoIP push — minimized / killed iOS, native CallKit —
        // is told the call is over through the SAME channel that raised it.
        // Without this, a backgrounded callee never learns the call ended
        // (its STOMP + Stream WS are both down) and the CallKit screen lingers
        // until Stream's ~ring-timeout. Async + swallowing; a no-op/404 when
        // the call was never created on Stream (all callees online).
        streamVideo.endCall(c.getStreamCallCid(), c.getCallerId());
    }

}

File: app/Features/Chat/Services/ChatCallService.php
<?php

namespace App\Features\Chat\Services;

use App\Features\Chat\Models\CallStatus;
use App\Features\Chat\Models\ChatCall;
use App\Features\Chat\Models\ChatCallParticipant;
use App\Features\Chat\Models\ParticipantStatus;
use App\Features\Chat\Models\PresenceStatus;
use App\Support\Exceptions\BadRequestException;
use App\Support\Exceptions\NotFoundException;
use App\Support\Pagination\PageQuery;
use Illuminate\Contracts\Pagination\LengthAwarePaginator;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

/**
 * Port of the Spring ChatCallService. Owns the call state machine
 * (RINGING → ANSWERED → ENDED/MISSED/REJECTED), the busy-signal gate, the
 * Stream VoIP ring/cancel side-effects, and presence BUSY toggling.
 */
class ChatCallService
{
    /** @var array<int, string> */
    private const OPEN_CALL_STATUSES = [CallStatus::RINGING, CallStatus::ANSWERED];

    /** @var array<int, string> */
    private const OPEN_PARTICIPANT_STATUSES = [ParticipantStatus::RINGING, ParticipantStatus::ANSWERED];

    /** @var ConversationService */
    private $conversations;

    /** @var PresenceService */
    private $presence;

    /** @var StreamTokenService */
    private $streamTokens;

    /** @var StreamVideoService */
    private $streamVideo;

    public function __construct(
        ConversationService $conversations,
        PresenceService $presence,
        StreamTokenService $streamTokens,
        StreamVideoService $streamVideo
    ) {
        $this->conversations = $conversations;
        $this->presence = $presence;
        $this->streamTokens = $streamTokens;
        $this->streamVideo = $streamVideo;
    }

    private function ringTimeoutSeconds(): int
    {
        $v = (int) config('erp.chat.call.ring_timeout_seconds');
        return $v <= 0 ? 60 : $v;
    }

    private function acceptGraceSeconds(): int
    {
        $v = (int) config('erp.chat.call.accept_grace_seconds');
        return $v < 0 ? 5 : $v;
    }

    public function getById(int $callId): ChatCall
    {
        $c = ChatCall::query()->with('participants')->find($callId);
        if (! $c) {
            throw new NotFoundException('Call not found');
        }
        return $c;
    }

    public function historyForUser(int $userId, PageQuery $query): LengthAwarePaginator
    {
        return ChatCall::query()
            ->whereIn('id', function ($q) use ($userId) {
                $q->select('call_id')->from('chat_call_participants')->where('user_id', $userId);
            })
            ->with('participants')
            ->orderBy('started_at', 'desc')
            ->paginate($query->pageSize, ['*'], 'page', $query->page);
    }

    public function historyForConversation(int $convId, int $userId, PageQuery $query): LengthAwarePaginator
    {
        $this->conversations->requireMember($convId, $userId);
        return ChatCall::query()
            ->where('conversation_id', $convId)
            ->with('participants')
            ->orderBy('started_at', 'desc')
            ->paginate($query->pageSize, ['*'], 'page', $query->page);
    }

    /**
     * @param array<string, mixed> $data  type
     */
    public function start(int $convId, int $callerId, array $data): ChatCall
    {
        $this->conversations->requireMember($convId, $callerId);
        if ($this->existsActiveCallForUser($callerId)) {
            throw new BadRequestException('Caller is already in an active call');
        }

        $call = DB::transaction(function () use ($convId, $callerId, $data) {
            $c = new ChatCall();
            $c->conversation_id = $convId;
            $c->caller_id = $callerId;
            $c->type = $data['type'];
            $c->status = CallStatus::RINGING;
            $c->started_at = Carbon::now();
            $c->save();

            // Now that the row has an id, stamp the Stream Video CID so every
            // participant joins the same Stream call.
            $c->stream_call_cid = $this->streamTokens->cidForCall((int) $c->id);
            $c->save();

            $memberIds = $this->conversations->memberUserIds($convId);
            foreach ($memberIds as $uid) {
                $uid = (int) $uid;
                $p = new ChatCallParticipant();
                $p->call_id = $c->id;
                $p->user_id = $uid;
                if ($uid === $callerId) {
                    $p->status = ParticipantStatus::ANSWERED;
                    $p->joined_at = Carbon::now();
                } else {
                    $p->status = ParticipantStatus::RINGING;
                }
                $p->save();
            }

            return $c;
        });

        $memberIds = $this->conversations->memberUserIds($convId);

        // Caller is busy from the moment the call starts.
        $this->presence->markBusy($callerId);

        // Ring the callees server-side via Stream's VoIP push — but ONLY for
        // callees who are OFFLINE (no live session, i.e. backgrounded/killed).
        // ONLINE callees already got the in-app overlay from the STOMP/WS
        // call.invite, so re-ringing them via VoIP would double-ring and
        // contest the audio session. OFFLINE callees still get the VoIP push.
        $ringTargets = [];
        foreach ($memberIds as $uid) {
            $uid = (int) $uid;
            if ($uid === $callerId) {
                continue;
            }
            if ($this->presence->statusOf($uid) === PresenceStatus::OFFLINE) {
                $ringTargets[] = $uid;
            }
        }
        if (empty($ringTargets)) {
            Log::info('[call] callId='.$call->id.' — all callees ONLINE; skipping Stream VoIP ring '
                .'(in-app overlay handles foreground, no CallKit)');
        } else {
            // Include the caller so Stream makes them the creator (creators are
            // NOT rung); only the OFFLINE callees in this set receive the push.
            $ringMembers = $ringTargets;
            $ringMembers[] = $callerId;
            Log::info('[call] callId='.$call->id.' — ringing OFFLINE callees via VoIP push: ['
                .implode(',', $ringTargets).']');
            $this->streamVideo->ring($call->stream_call_cid, $callerId, $ringMembers);
        }

        return $this->getById((int) $call->id);
    }

    public function accept(int $callId, int $userId): ChatCall
    {
        return DB::transaction(function () use ($callId, $userId) {
            $c = $this->getById($callId);

            // Grace-window revival — if the sweeper just marked it MISSED but the
            // accept arrives within acceptGraceSeconds, restore RINGING and proceed.
            if ($c->status === CallStatus::MISSED) {
                $ageSec = $c->started_at->diffInSeconds(Carbon::now());
                $graceCutoff = $this->ringTimeoutSeconds() + $this->acceptGraceSeconds();
                if ($ageSec <= $graceCutoff) {
                    Log::info('[call] REVIVE callId='.$callId.' accepter user='.$userId
                        .' ageSec='.$ageSec.' graceCutoff='.$graceCutoff);
                    $c->status = CallStatus::RINGING;
                    $c->ended_at = null;
                    $c->end_reason = null;
                    $c->duration_seconds = null;
                    $c->save();
                    $me = $this->participant($callId, $userId);
                    if ($me->status === ParticipantStatus::MISSED) {
                        $me->status = ParticipantStatus::RINGING;
                        $me->left_at = null;
                        $me->save();
                    }
                } else {
                    throw new BadRequestException('Call already ended');
                }
            } elseif ($c->status !== CallStatus::RINGING && $c->status !== CallStatus::ANSWERED) {
                throw new BadRequestException('Call already ended');
            }

            $p = $this->participant($callId, $userId);
            if ($p->status !== ParticipantStatus::RINGING) {
                return $this->getById($callId);
            }
            $p->status = ParticipantStatus::ANSWERED;
            $p->joined_at = Carbon::now();
            $p->save();

            if ($c->status === CallStatus::RINGING) {
                $c->status = CallStatus::ANSWERED;
                $c->answered_at = Carbon::now();
                $c->save();
            }
            $this->presence->markBusy($userId);

            return $this->getById($callId);
        });
    }

    public function reject(int $callId, int $userId, ?string $reason): ChatCall
    {
        return DB::transaction(function () use ($callId, $userId, $reason) {
            $c = $this->getById($callId);
            $p = $this->participant($callId, $userId);
            if ($p->status === ParticipantStatus::RINGING) {
                $p->status = ParticipantStatus::REJECTED;
                $p->left_at = Carbon::now();
                $p->save();
            }
            // If everyone else (other than caller) has rejected/left, the call ends.
            $anyoneActive = $c->participants()
                ->where('user_id', '<>', $c->caller_id)
                ->whereIn('status', [ParticipantStatus::RINGING, ParticipantStatus::ANSWERED])
                ->exists();
            if (! $anyoneActive && $c->status === CallStatus::RINGING) {
                $this->endCallInternal(
                    $c,
                    $c->status === CallStatus::RINGING ? CallStatus::REJECTED : CallStatus::ENDED,
                    $reason !== null ? $reason : 'rejected'
                );
            }
            return $this->getById($callId);
        });
    }

    public function hangup(int $callId, int $userId): ChatCall
    {
        return DB::transaction(function () use ($callId, $userId) {
            $c = $this->getById($callId);
            $p = $this->participant($callId, $userId);
            if ($p->status === ParticipantStatus::ANSWERED || $p->status === ParticipantStatus::RINGING) {
                $p->status = ParticipantStatus::LEFT;
                $p->left_at = Carbon::now();
                $p->save();
            }

            // Caller leaving ends the call for everyone.
            if ($userId === (int) $c->caller_id) {
                $this->endCallInternal($c, CallStatus::ENDED, 'caller_left');
                return $this->getById($callId);
            }

            // Otherwise: if no other callee is still in, end it.
            $anyoneStillActive = $c->participants()
                ->where('user_id', '<>', $c->caller_id)
                ->whereIn('status', [ParticipantStatus::RINGING, ParticipantStatus::ANSWERED])
                ->exists();
            if (! $anyoneStillActive && $c->status !== CallStatus::ENDED) {
                $this->endCallInternal($c, CallStatus::ENDED, 'all_callees_left');
            } else {
                // Call continues; the user who left is no longer BUSY.
                $this->presence->clearBusy($userId);
            }
            return $this->getById($callId);
        });
    }

    /**
     * Sweep every RINGING call older than the configured ring timeout and
     * transition it to MISSED. Returns the freshly-ended calls (reloaded with
     * participants) so the caller can fan out STOMP + FCM notifications.
     *
     * @return array<int, ChatCall>
     */
    public function sweepStaleRinging(): array
    {
        $cutoff = Carbon::now()->subSeconds($this->ringTimeoutSeconds());

        $staleIds = ChatCall::query()
            ->where('status', CallStatus::RINGING)
            ->where('started_at', '<', $cutoff)
            ->pluck('id')
            ->all();

        $ended = [];
        foreach ($staleIds as $id) {
            $id = (int) $id;
            DB::transaction(function () use ($id) {
                $c = $this->getById($id);
                if ($c->status !== CallStatus::RINGING) {
                    return;
                }
                foreach ($c->participants as $p) {
                    if ($p->status === ParticipantStatus::RINGING) {
                        $p->status = ParticipantStatus::MISSED;
                        $p->left_at = Carbon::now();
                        $p->save();
                    }
                }
                $this->endCallInternal($c, CallStatus::MISSED, 'no_answer');
                Log::info('[call] AUTO-MISSED callId='.$c->id.' (timeout='.$this->ringTimeoutSeconds().'s)');
            });
            $ended[] = $this->getById($id);
        }
        return $ended;
    }

    private function endCallInternal(ChatCall $c, string $status, string $reason): void
    {
        $c->status = $status;
        $c->ended_at = Carbon::now();
        $c->end_reason = $reason;
        if ($c->answered_at !== null) {
            $c->duration_seconds = (int) $c->answered_at->diffInSeconds($c->ended_at);
        } else {
            $c->duration_seconds = 0;
        }
        $c->save();

        // Clear BUSY for everyone who was active in this call. Read fresh from
        // the DB so any participant status mutated earlier in this same flow
        // (e.g. the caller flipped to LEFT in hangup) is reflected.
        foreach ($c->participants()->get() as $p) {
            if ($p->status === ParticipantStatus::ANSWERED || $p->status === ParticipantStatus::LEFT) {
                $this->presence->clearBusy((int) $p->user_id);
            }
        }
        // Caller, even if they never "answered", was busy from start.
        if ($c->caller_id !== null) {
            $this->presence->clearBusy((int) $c->caller_id);
        }

        // Cancel the Stream ring for every member — the single chokepoint for
        // every terminal end, so a still-ringing (backgrounded/killed) callee is
        // told the call is over through the same VoIP channel that raised it.
        $this->streamVideo->endCall($c->stream_call_cid, $c->caller_id !== null ? (int) $c->caller_id : 0);
    }

    private function participant(int $callId, int $userId): ChatCallParticipant
    {
        $p = ChatCallParticipant::query()
            ->where('call_id', $callId)->where('user_id', $userId)->first();
        if (! $p) {
            throw new NotFoundException('You are not a participant in this call');
        }
        return $p;
    }

    /** Busy-signal check: is this user mid-call right now? */
    private function existsActiveCallForUser(int $userId): bool
    {
        return ChatCall::query()
            ->whereIn('status', self::OPEN_CALL_STATUSES)
            ->whereExists(function ($q) use ($userId) {
                $q->selectRaw('1')->from('chat_call_participants as p')
                    ->whereColumn('p.call_id', 'chat_calls.id')
                    ->where('p.user_id', $userId)
                    ->whereIn('p.status', self::OPEN_PARTICIPANT_STATUSES);
            })
            ->exists();
    }
}
